Enables ICMP network policy function by default

chez-shanpu opened the pull request:
resolves: #14609

Previously, ICMP network policy function is disabled by default because of kernel complexity issue (https://github.com/cilium/cilium/pull/16516).
However, currently this issue is resolved, and this PR makes ICMP policy function is enabled by default.
Also this PR contains e2e tests and documents for ICMP rules .

chez-shanpu committed:
docs: update documents for ICMP CNP

Add the section for ICMP/ICMPv6 policy in the "Layer 4 Eexamples" chapter.

Signed-off-by: Tomoki Sugiura <tomoki.sugiura@mail.shanpu.info>

chez-shanpu committed:
test: add tests for ICMP CNP

Add e2e test for ICMP Cilium Network Policies.

Signed-off-by: Tomoki Sugiura <tomoki.sugiura@mail.shanpu.info>

pchaigno commented on daemon/cmd/daemon_main.go:
```diff
@@ -1094,10 +1094,6 @@ func initializeFlags() {
 
-	flags.Bool(option.EnableICMPRules, false, "Enable ICMP-based rule support for Cilium Network Policies")
-	flags.MarkHidden(option.EnableICMPRules)
-	option.BindEnv(option.EnableICMPRules)
```
Given this is still in beta support, I think we should keep the flag, in case we need to disable it quickly to workaround some bug.

chez-shanpu replied:
Thank you for your comment!
Would it mean that keep this flag and make it hidden state and set "true" by default?

pchaigno replied:
Yes, I think that would be a good approach. The pros of keeping it hidden is that we can remove it once we think things are stable enough. I'd rather not maintain such a flag in the long term because it's one more option we have to cover in complexity tests.

chez-shanpu replied:
I see. I've fixed codes and set "true" by default :)

pchaigno commented on test/k8s/net_policies.go:
```diff
@@ -370,6 +378,147 @@ var _ = SkipDescribeIf(func() bool {
+			_, err = kubectl.CiliumPolicyAction(
+				namespaceForTest, ICMPv6PolicyDeny,
+				helpers.KubectlDelete, helpers.HelperTimeout)
+			Expect(err).Should(BeNil(), "Cannot delete ICMPv6 Policy Deny")
```
If the test fails on one of the previous commands, the ICMP policies won't be deleted, right? That is likely to cause issues in subsequent tests, so we should delete them in an AfterAll step or similar.

chez-shanpu replied:
In my understanding, all of the network policy resources are deleted by `kubectl` at the AfterEach() (L241~244) whenever It() ends, so I think we don't have to add codes for cleaning ICMP policies. But It cannot be enough?

pchaigno replied:
Ah, nice! Yes, nothing else needed.

pchaigno reviewed (approved):
Thanks for the quick turnaround! A couple comments below but looks good overall! :crossed_fingers: this can make it into the v1.12 release notes.

chez-shanpu committed:
Make enable-icmp-rules flag is true by default

This commit makes `enable-icmp-rules` feature-flag is true and enables ICMP/ICMPv6 network policy rules by default.
(Previously ICMP policy rules function is disabled by default because of kernel complexity issue.)

Signed-off-by: Tomoki Sugiura <tomoki.sugiura@mail.shanpu.info>

ldelossa reviewed (approved):
LGTM

nathanjsweet reviewed (approved):
Related API work looks good.

qmonnet commented on Documentation/policy/language.rst:
```diff
@@ -731,6 +731,57 @@ ports other than port 80.
+   .. tabs::
+     .. group-tab:: k8s YAML
+
+        .. literalinclude:: ../../examples/policies/l4/icmp.yaml
```
You can have syntax highlighting on the files that you include, too

```suggestion
        .. literalinclude:: ../../examples/policies/l4/icmp.yaml
           :language: yaml

```

chez-shanpu replied:
Thank you! I've updated :)

qmonnet commented on Documentation/policy/language.rst:
```diff
@@ -731,6 +731,57 @@ ports other than port 80.
+        .. literalinclude:: ../../examples/policies/l4/icmp.yaml
+     .. group-tab:: JSON
+
+        .. literalinclude:: ../../examples/policies/l4/icmp.json
```
```suggestion
        .. literalinclude:: ../../examples/policies/l4/icmp.json
           :language: json
```

qmonnet commented on Documentation/policy/language.rst:
```diff
@@ -731,6 +731,57 @@ ports other than port 80.
+.. only:: epub or latex
+
+        .. literalinclude:: ../../examples/policies/l4/icmp.json
+
```
```suggestion
        .. literalinclude:: ../../examples/policies/l4/icmp.json
           :language: json
```
